source/solvers/cahn_hilliard.cc

```diff
@@ -39,52 +40,210 @@ CahnHilliard<dim>::setup_assemblers()
+                      &source_term);
+
+  const DoFHandler<dim> *dof_handler_fluid =
+    multiphysics->get_dof_handler(PhysicsID::fluid_dynamics);
+
+  typename DoFHandler<dim>::active_cell_iterator velocity_cell(
+    &(*triangulation), cell->level(), cell->index(), dof_handler_fluid);
+
+  if (multiphysics->fluid_dynamics_is_block())
+    {
+      scratch_data.reinit_velocity(velocity_cell,
+                                   *multiphysics->get_block_solution(
+                                     PhysicsID::fluid_dynamics));
+    }
+  else
+    {
+      scratch_data.reinit_velocity(
+        velocity_cell, *multiphysics->get_solution(PhysicsID::fluid_dynamics));
+    }
+
+  scratch_data.calculate_physical_properties();
+
+  copy_data.reset();
+
+  for (auto &assembler : this->assemblers)
+    {
+      assembler->assemble_matrix(scratch_data, copy_data);
+    }
+
+
+  cell->get_dof_indices(copy_data.local_dof_indices);
 }
 
 template <int dim>
 void
-CahnHilliard<dim>::copy_local_matrix_to_global_matrix()
+CahnHilliard<dim>::copy_local_matrix_to_global_matrix(
+  const StabilizedMethodsCopyData &copy_data)
 {
-  return;
+  if (!copy_data.cell_is_local)
+    return;
+
+  const AffineConstraints<double> &constraints_used = this->zero_constraints;
+  constraints_used.distribute_local_to_global(copy_data.local_matrix,
+                                              copy_data.local_dof_indices,
+                                              system_matrix);
 }
 
 
 template <int dim>
 void
 CahnHilliard<dim>::assemble_system_rhs()
 {
-  return;
+  // TimerOutput::Scope t(this->computing_timer, "Assemble RHS");
+  this->system_rhs = 0;
+  setup_assemblers();
+
+  const DoFHandler<dim> *dof_handler_fluid =
+    multiphysics->get_dof_handler(PhysicsID::fluid_dynamics);
+
+  auto scratch_data = CahnHilliardScratchData<dim>(
+    this->simulation_parameters.physical_properties_manager,
+    *this->fe,
+    *this->cell_quadrature,
+    *this->mapping,
+    dof_handler_fluid->get_fe(),
+    *this->face_quadrature);
+
+  WorkStream::run(this->dof_handler.begin_active(),
+                  this->dof_handler.end(),
+                  *this,
+                  &CahnHilliard::assemble_local_system_rhs,
+                  &CahnHilliard::copy_local_rhs_to_global_rhs,
+                  scratch_data,
+                  StabilizedMethodsCopyData(this->fe->n_dofs_per_cell(),
+                                            this->cell_quadrature->size()));
+
+  this->system_rhs.compress(VectorOperation::add);
 }
 
 template <int dim>
 void
-CahnHilliard<dim>::assemble_local_system_rhs()
+CahnHilliard<dim>::assemble_local_system_rhs(
+  const typename DoFHandler<dim>::active_cell_iterator &cell,
+  CahnHilliardScratchData<dim> &                        scratch_data,
+  StabilizedMethodsCopyData &                           copy_data)
 {
-  return;
+  copy_data.cell_is_local = cell->is_locally_owned();
+  if (!cell->is_locally_owned())
```

Comment:
```suggestion
  if (!copy_data.cell_is_local)
```

Reply:
Could be all combined in the if I believe.

Reply:
I need to update the copy_data.cell_is_local attribute, that's why I had two separate lines